cluster/gce/gci/configure-helper-testcase.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gci
+
+import (
+	"github.com/golang/glog"
+
+	goruntime "runtime"
+
+	"testing"
+
+	"text/template"
+
+	"io/ioutil"
+
+	"k8s.io/api/core/v1"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"os"
+	"path/filepath"
+
+	"fmt"
+	"io"
+	"os/exec"
+	"path"
+
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+)
+
+const (
+	envScriptFileName         = "kube-env"
+	configureHelperScriptName = "configure-helper.sh"
+)
+
+type ManifestTestCase struct {
+	pod                 v1.Pod
+	envScriptPath       string
+	manifest            string
+	auxManifests        []string
+	kubeHome            string
+	manifestSources     string
+	manifestDestination string
+	manifestTemplate    string
+	manifestFuncName    string
+	manifestTemplateDir string
+	t                   *testing.T
+}
+
+func NewManifestTestCase(manifestUnderTest, funcUnderTest string, auxManifests []string, t *testing.T) *ManifestTestCase {
+	c := &ManifestTestCase{}
+	c.t = t
+	d, err := ioutil.TempDir("/tmp", "")
+	if err != nil {
+		c.t.Fatalf("Failed to create temp directory: %v", err)
+	}
+
+	c.kubeHome = d
+	glog.V(3).Infof("KUBE_HOME of the test case is %s", c.kubeHome)
+
+	c.manifestFuncName = funcUnderTest
+	glog.V(3).Infof("configure-helper.sh function under test is: %s", c.manifestFuncName)
+
+	c.manifest = manifestUnderTest
+	glog.V(3).Infof("Manifest under test is %s", manifestUnderTest)
+
+	// Some manifests consume other manifests. For example, kube-apiserver.manifest may consume kms-plugin-container.manifest.
+	c.auxManifests = auxManifests
+	glog.V(3).Infof("Auxiliary manifests: %v", c.auxManifests)
+
+	c.envScriptPath = filepath.Join(c.kubeHome, envScriptFileName)
+	glog.V(3).Infof("kube-env variables will be sourced by bash from %s", c.envScriptPath)
+
+	c.manifestSources = path.Join(filepath.Join(c.kubeHome, "kube-manifests", "kubernetes", "gci-trusty"))
+	glog.V(3).Infof("Function under test will access the manifest from %s", c.manifestSources)
+
+	// full path of this go file
+	_, testFilePath, _, _ := goruntime.Caller(0)
+	c.manifestTemplateDir = filepath.Join(path.Dir(path.Dir(testFilePath)), "manifests")
```

Comment:
filepath.Dir, in both cases. Ideally this file should only import "path/filepath" and not "path"

Another question is: if you only read/exec files in "../manifests/", why not put this test code there?

Reply:
Removed reference to path.

I feel that I am not testing manifests (data), but configure-helper.sh. Typically, we would expect tests to be close to their targets.